sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/ImplementISerializableCorrectly.cs

```diff
@@ -0,0 +1,246 @@
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(rule);
+
+        protected override void Initialize(SonarAnalysisContext context)
+        {
+            context.RegisterSyntaxNodeActionInNonGenerated(c =>
+            {
+                if (c.SemanticModel.Compilation.IsTest())
+                {
+                    return;
+                }
+
+                var classDeclaration = (ClassDeclarationSyntax)c.Node;
+                var typeSymbol = c.SemanticModel.GetDeclaredSymbol(classDeclaration);
+
+                if (!ImplementsISerializable(typeSymbol))
+                {
+                    return;
+                }
+
+                var getObjectData = typeSymbol.GetMembers()
+                    .OfType<IMethodSymbol>()
+                    .FirstOrDefault(KnownMethods.IsGetObjectData);
+
+                var implementationErrors = new List<SecondaryLocation>();
+
+                implementationErrors.AddRange(CheckSerializableAttribute(classDeclaration, typeSymbol));
+                implementationErrors.AddRange(CheckConstructor(classDeclaration, typeSymbol));
+                implementationErrors.AddRange(CheckGetObjectDataAccessibility(typeSymbol, getObjectData));
+                implementationErrors.AddRange(CheckGetObjectData(typeSymbol, getObjectData, c.SemanticModel, classDeclaration));
+
+                if (implementationErrors.Count > 0)
+                {
+                    c.ReportDiagnostic(Diagnostic.Create(rule, classDeclaration.Identifier.GetLocation(),
+                        additionalLocations: implementationErrors.ToAdditionalLocations(),
+                        properties: implementationErrors.ToProperties()));
+                }
+            },
+            SyntaxKind.ClassDeclaration);
+        }
+
+        private static IEnumerable<SecondaryLocation> CheckSerializableAttribute(ClassDeclarationSyntax classDeclaration, INamedTypeSymbol typeSymbol)
+        {
+            if (!HasSerializableAttribute(typeSymbol) &&
+                !typeSymbol.IsAbstract)
+            {
+                yield return new SecondaryLocation(classDeclaration.Keyword.GetLocation(),
+                    $"Add 'System.SerializableAttribute' attribute on '{typeSymbol.Name}' because it implements 'ISerializable'.");
+            }
+        }
+
+        private static IEnumerable<TSyntax> GetDeclarations<TSyntax>(ISymbol symbol)
+        {
+            if (symbol == null)
+            {
+                return Enumerable.Empty<TSyntax>();
+            }
+
+            return symbol.DeclaringSyntaxReferences.Select(r => r.GetSyntax()).Cast<TSyntax>();
+        }
+
+        private static IEnumerable<SecondaryLocation> CheckGetObjectDataAccessibility(INamedTypeSymbol typeSymbol, IMethodSymbol getObjectData)
+        {
+            if (!typeSymbol.IsSealed &&

```

Comment:
You are using a lot of not in this `if`, it would probably be easier to invert and return `Enumerable.Empty<SecondaryLocaiton>()`